src/Autofac/Core/Registration/ComponentRegistryBuilder.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+using Autofac.Builder;
+using Autofac.Util;
+
+namespace Autofac.Core.Registration
+{
+    /// <summary>
+    /// Used to build a <see cref="IComponentRegistry" />.
+    /// </summary>
+    internal class ComponentRegistryBuilder : Disposable, IComponentRegistryBuilder
+    {
+        /// <summary>
+        /// Protects instance variables from concurrent access.
+        /// </summary>
+        private readonly object _synchRoot = new object();
```

Comment:
Really amazing, first-class PR :-) !!! 👏 🎉 

Just one (late) question - does this class need synchronization now? The only times registry builders are mutated are during initial container build-up (required to be single-threaded) and during scope creation (also single-threaded).

Reply:
Thank you :-).
Now that I look at this again, I think you are right and this can be removed. I don't remember why I left it in, or if I actually had a reason to leave it in.

Reply:
Thanks for the reply! @alexmg @tillig I'll PR some changes through around this and `Properties` if it sounds sane to you both :-)

Reply:
Sounds great! This is all in the pr-immutable-container branch. I still haven't had a chance to really dig in due to some personal stuff coming up unexpectedly.